File: problems/0010_regular-expression-matching/bottom-to-top/solution.rb
#!/usr/bin/ruby

# @param {String} s
# @param {String} p
# @return {Boolean}
def is_match(s, p)
    memo = Array.new(s.length + 1) {
        Array.new(p.length + 1, false)
    }
    memo[s.length][p.length] = true
    for i in (s.length).downto(0)
        for j in (p.length - 1).downto(0)
            match_char = (i < s.length and
                    (p[j] == s[i] or p[j] == '.'))
            if j < p.length - 1 and p[j + 1] == '*'
                memo[i][j] = (memo[i][j + 2] or (match_char and memo[i + 1][j]))
            elsif
                memo[i][j] = (match_char and memo[i + 1][j + 1])
            end
        end
    end
    return memo[0][0]
end

File: problems/0010_regular-expression-matching/bottom-to-top/Solution.kt
class Solution {
    fun isMatch(s: String, p: String): Boolean {
        val memo = Array(s.length + 1) { Array(p.length + 1) { false } }
        memo[s.length][p.length] = true
        for (i in s.length downTo 0)
            for (j in p.length - 1 downTo 0) {
                val matchChar = (i < s.length &&
                        (p[j] == s[i] || p[j] == '.'))
                memo[i][j] = if (j < p.length - 1 && p[j + 1] == '*')
                            (memo[i][j + 2] || (matchChar && memo[i + 1][j]))
                        else
                            (matchChar && memo[i + 1][j + 1])
            }
        return memo[0][0]
    }
}
